File: bin/native/bend2/effs/tcp_accept.js
// TCP
// ===

// The loop parked the request until the listener was readable; an accept
// that still finds no connection (the listener is non-blocking) parks
// again. The accepted socket is non-blocking for life.
function tcp_accept(listener, k) {
  const sys = io_sys();
  const lfd = listener;
  const go = () => {
    const fd = sys.accept(lfd, null, null);
    if (fd < 0) {
      const code = sys.errno();
      if (code === (sys.mac ? 35 : 11)) {
        io_park_on(lfd, false, k, go);
        return undefined;
      }
      return io_tup(listener, io_fail(code));
    }
    if (sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800)) < 0) {
      const code = sys.errno();
      sys.close(fd);
      return io_tup(listener, io_fail(code));
    }
    return io_tup(listener, io_done(fd));
  };
  return go();
}

function tcp_accept_need() {
  return { read: true };
}

File: bin/native/bend2/effs/tcp_accept.c
// TCP
// ===

// The loop parked the request until the listener was readable; an accept
// that still finds no connection (the listener is non-blocking) parks
// again. The accepted socket is non-blocking for life.
static Term tcp_accept_more(Env e, IoWork* w) {
  int fd  = (int)w->hand;
  int got = accept(fd, NULL, NULL);
  if (got >= 0 && fcntl(got, F_SETFL, fcntl(got, F_GETFL) | O_NONBLOCK) < 0) {
    close(got);
    got = -1;
  }
  io_sys_end(w, got);
  if (w->code == EAGAIN) {
    return io_wait_on(w, fd, POLLIN, 0, tcp_accept_more);
  }
  Term r = w->code != 0 ? io_fail(e, w->code, NULL) : io_done(e, io_hand(got));
  return io_tup(e, io_hand(fd), r);
}

Term tcp_accept_run(Env e, Term* f, IoWork* w) {
  w->hand = (intptr_t)io_hand_v(f[0]);
  return tcp_accept_more(e, w);
}

static void __attribute__((constructor)) tcp_accept_use(void) {
  io_eff(CID_TCP_ACCEPT, tcp_accept_run, IO_READ);
}
